easybuild/main.py

```diff
@@ -213,162 +214,100 @@ def clean_exit(logfile, tmpdir, testing, silent=False):
-    # So emulate this here to allow (module) scripts depending on that to work
-    if '_' not in os.environ:
-        os.environ['_'] = sys.executable
-
-    # purposely session state very early, to avoid modules loaded by EasyBuild meddling in
-    init_session_state = session_state()
-
-    eb_go, cfg_settings = set_up_configuration(args=args, logfile=logfile, testing=testing)
-    options, orig_paths = eb_go.options, eb_go.args
-
-    global _log
-    (build_specs, _log, logfile, robot_path, search_query, eb_tmpdir, try_to_generate,
-     from_pr_list, tweaked_ecs_paths) = cfg_settings
-
-    # load hook implementations (if any)
-    hooks = load_hooks(options.hooks)
-
-    run_hook(START, hooks)
-
-    if modtool is None:
+    orig_paths, opts_per_ec = parse_easystack(easystack_path)
+
+    _log.debug("Start build loop over items in the EasyStack file: %s" % orig_paths)
+    # TODO: insert fast loop that validates if all command line options are valid. If there are errors in options,
+    # we want to know early on, and this loop potentially builds a lot of packages and could take very long
+    # for path in orig_paths:
+    #     validate_command_opts(args, opts_per_ec[path])
+
+    # Loop over each item in the EasyStack file, each time updating the config
+    # This is because each item in an EasyStack file can have options associated with it
+    do_cleanup = True
+    for path in orig_paths:
+        _log.debug("Starting build for %s" % path)
+        # Whipe easyconfig caches
+        easyconfig._easyconfigs_cache.clear()
+        easyconfig._easyconfig_files_cache.clear()
+
+        # If EasyConfig specific arguments were supplied in EasyStack file
+        # merge arguments with original command line args
+        if path in opts_per_ec:
+            _log.debug("EasyConfig specific options have been specified for "
+                       "%s in the EasyStack file: %s" % (path, opts_per_ec[path]))
+            if args is None:
+                args = sys.argv[1:]
+            ec_args = dict_to_argslist(opts_per_ec[path])
+            # By appending ec_args to args, ec_args take priority
+            new_args = args + ec_args
+            _log.info("Argument list for %s after merging command line arguments with EasyConfig specific "
+                      "options from the EasyStack file: %s" % (path, new_args))
+        else:
+            # If no EasyConfig specific arguments are defined, use original args.
+            # That way,set_up_configuration restores the original config
+            new_args = args
+
+        # Reconfigure
+        eb_go, cfg_settings = set_up_configuration(args=new_args, logfile=logfile, testing=testing,
+                                                   reconfigure=True)
+        # Since we reconfigure, we should also reload hooks and get current module tools
+        hooks = load_hooks(eb_go.options.hooks)
         modtool = modules_tool(testing=testing)
```

Comment:
@boegel I figured that we should also re-fetch the modules-tool (didn't do that before) right after `set_up_configuration`, agreed? It could be that different items in the EasyStack file use different module tools. From what I understand, `modules_tool` fetches this from `ConfigurationVariables` (see [here](https://github.com/casparvl/easybuild-framework/blob/5096c4cc324bbd9642de7070cb87cc672a7d3bc0/easybuild/tools/config.py#L704)), which is reset by running `set_up_configuration` with `reconfigure=True`. Correct?

Reply:
Makes sense, yes, since we're passing this down to `process_eb_args`